### DISCUSSION TITLE
PYNQ-ZU support with VivadoAccelerator backend

### DISCUSSION THREAD

**Discussion by jacopoabramo** [fastmachinelearning/hls4ml#633] [General]
Greetings,

as requested in issue #623 I tested the integration of the PYNQ-ZU board. I managed to create a bitstream of the model presented in the deployment tutorial, and it works fine on the PYNQ-ZU.

A couple of points related to the port itself:

- [the part number](https://github.com/fastmachinelearning/hls4ml/blob/5cf197e401fb7cf912693566787221eb112a5b32/hls4ml/templates/vivado_accelerator/pynq-zu/tcl_scripts/axi_stream_design.tcl#L5) is incorrect, it should be `xczu5eg-sfvc784-1-e`
- The final accuracy of the model inferred on the FPGA is around 16% less than the original model (74% QKeras model, 58% on FPGA) with these settings:

```python
hls4ml.model.optimizer.passes.qkeras.OutputRoundingSaturationMode.layers = ['Activation']
hls4ml.model.optimizer.passes.qkeras.OutputRoundingSaturationMode.rounding_mode = 'AP_RND'
hls4ml.model.optimizer.passes.qkeras.OutputRoundingSaturationMode.saturation_mode = 'AP_SAT'

config = hls4ml.utils.config_from_keras_model(model, granularity='name')
config['LayerName']['softmax']['exp_table_t'] = 'ap_fixed<18,8>'
config['LayerName']['softmax']['inv_table_t'] = 'ap_fixed<18,4>'
for layer in ['fc1', 'fc2', 'fc3', 'output']:
    config['LayerName'][layer]['ReuseFactor'] = 64
```

I assume though that changing some conversion strategy to increase the accuracy the drop should reduce significantly.

**Comment by jacopoabramo** [comment]
Some issues I encountered unrelated to hls4ml but related to Vivado:

- due to the [revision overflow bug](https://support.xilinx.com/s/article/76960?language=en_US) affected by versions up to 2021.2, it would be nice to notify users to adapt their Vivado versions with the patch they provide somewhere in the package documentation;
- it happened to me that for some reason when launching Vivado in batch mode the repository paths are not set, hence the pynq-zu board files were not found. I fixed this by adding a Vivado_init.tcl script which would set the repository paths with the command `set_param board.repoPaths {paths to board files}`. This script goes in `<installation_folder>/Xilinx/Vivado/<vivado_version>/scripts/`
